core/src/main/java/io/grpc/util/RoundRobinLoadBalancerFactory.java

```diff
@@ -296,58 +316,69 @@ private ConnectivityState getAggregatedState() {
+       * Returns the subchannel associated to the stickiness value if available in both the
        * registry and the round robin list, otherwise associates the given subchannel with the
        * stickiness key in the registry and returns the given subchannel.
        */
       @Nonnull
-      synchronized Subchannel maybeRegister(
+      Subchannel maybeRegister(
           String stickinessValue, @Nonnull Subchannel subchannel, List<Subchannel> rrList) {
-        Subchannel existingSubchannel = getSubchannel(stickinessValue);
-        if (existingSubchannel != null && rrList.contains(existingSubchannel)) {
-          return existingSubchannel;
+        final Ref<Subchannel> newSubchannelRef = subchannel.getAttributes().get(STICKY_REF);
+        Ref<Subchannel> existingSubchannelRef = null;
+        while (true) {
+          if (existingSubchannelRef != null) {
+            if (stickinessMap.replace(stickinessValue, existingSubchannelRef, newSubchannelRef)) {
+              addToLruQueue(stickinessValue, newSubchannelRef, false);
+              return subchannel;
+            }
+            existingSubchannelRef = stickinessMap.get(stickinessValue);
+          } else if ((existingSubchannelRef =
+                  stickinessMap.putIfAbsent(stickinessValue, newSubchannelRef)) == null) {
+            addToLruQueue(stickinessValue, newSubchannelRef, true);
+            return subchannel;
+          }
+          if (existingSubchannelRef != null) {
+            Subchannel existingSubchannel = existingSubchannelRef.value;
+            if (existingSubchannel != null && rrList.contains(existingSubchannel)) {
+              return existingSubchannel;
+            }
+          }
         }
+      }
 
-        Ref<Subchannel> subchannelRef = subchannelRefs.get(subchannel);
-        if (subchannelRef == null) {
-          subchannelRef = new Ref<Subchannel>(subchannel);
-          subchannelRefs.put(subchannel, subchannelRef);
+      private void addToLruQueue(String value, Ref<Subchannel> subchannelRef, boolean pruneFirst) {
+        if (pruneFirst) {
+          Entry<String,Ref<Subchannel>> oldEntry;
+          while (stickinessMap.size() >= MAX_ENTRIES && (oldEntry = lruQueue.poll()) != null) {
+            stickinessMap.remove(oldEntry.getKey(), oldEntry.getValue());
+          }
         }
-        stickinessMap.put(stickinessValue, subchannelRef);
-        return subchannel;
+        lruQueue.add(new SimpleImmutableEntry<String,Ref<Subchannel>>(value, subchannelRef));
       }
 
       /**
        * Unregister the subchannel from StickinessState.
        */
-      synchronized void remove(Subchannel subchannel) {
-        if (subchannelRefs.containsKey(subchannel)) {
-          subchannelRefs.get(subchannel).value = null;
-          subchannelRefs.remove(subchannel);
-        }
+      void remove(Subchannel subchannel) {
+        subchannel.getAttributes().get(STICKY_REF).value = null;
```

Comment:
> The problem is that shutting down the subchannel does not shrink the map

This matches the current behaviour - it's already the case that no entries are removed from `stickinessMap` when the subchannel is shutdown. As far as I could see, the separate `subchannelRefs` map just serves the purpose of "interning" the `Ref`s pointing to a given subchannel. This allows the canonical `Ref` to be nulled when the subchannel is shutdown, presumably to ensure immediate GC eligibility.

However it seems kind of redundant to maintain an independent subchannel->X lookup map (where X is constant), when a reference to X can be added to the subchannel itself. This is basically exactly what's already done for `ConnectivityStateInfo` - a cumbersome alternative to that could have been to not use a mutable Ref attribute and instead maintain a separate `Subchannel`->`ConnectivityStateInfo` map.

The only difference in this regard is the case where a stickiness val in the map is re-assigned to a new subchannel after its previous one was shutdown, and before its mapping has been evicted. With the new logic, a stale entry will remain in the `lruQueue` (actually it's not lru, it's lr-created), and so the queue could temporarily grow bigger than the map, but it will naturally be pruned as new map entries are added.

Back to the original problem - though the behaviour might not be worse, I agree it's not ideal. I could not think of an alternative which wouldn't introduce extra cost and/or complexity (stickinessval -> Subchannel is many-to-one so the reverse one-to-many map would need to be maintained to efficiently clear the refs when the Subchannel is shut-down, or otherwise introduce some O(n) stuff). A guava cache with weak values would be just what's needed I think, but am I correct that most of guava is off-limits?

Reply:
Oh, I misunderstood the current code, my mistake. 

>This matches the current behaviour - it's already the case that no entries are removed from stickinessMap when the subchannel is shutdown.